lib/google/cloud/gemserver/authentication.rb

```diff
@@ -52,9 +56,46 @@ def can_modify?
           editors.each do |editor|
             return true if extract_account(editor) == user
           end
+          puts "You are either not authenticated with gcloud or lack access" \
+            " to the gemserver."
           false
         end
 
+        ##
+        # Generates an access token from a user authenticated by gcloud.
+        #
+        # @return [String]
+        def access_token
+          return unless can_modify?
+          scope = ["https://www.googleapis.com/auth/cloud-platform"]
+          if ENV["GOOGLE_APPLICATION_CREDENTIALS"]
+            auth = Google::Auth::ServiceAccountCredentials.make_creds(
+              json_key_io: File.open(ENV["GOOGLE_APPLICATION_CREDENTIALS"]),
+              scope: scope
+            )
+          else
+            auth = Google::Auth.get_application_default scope
+          end
+          auth.fetch_access_token!
+        end
+
+        ##
+        # Uses the tokeninfo API to validate the given access token.
+        #
+        # @param [String] token The token to be validated.
+        #
+        # @return [Boolean]
+        def validate_token token
+          apis_url = "https://www.googleapis.com"
+          tokeninfo_endpoint = "/oauth2/v1/tokeninfo?access_token=#{token}"
+          uri = URI.parse apis_url
+          http = Net::HTTP.new uri.host, uri.port
+          http.use_ssl = true
+          res = http.request Net::HTTP::Post.new(tokeninfo_endpoint)
+
+          res.code.to_i == 200
```

Comment:
Looking at this more closely, it looks like I didn't get this right either. The tokeninfo call checks the integrity of the token but not its access. Now I'm not sure how to check for certain access. We might need to use the access token to attempt some admin call (like appengine.services.get) and make sure it succeeds. Let's look at this again on Monday.